python/src/python-module.cc

```diff
@@ -0,0 +1,76 @@
+#include <Python.h>
+
+#include "internal/eval.hh"
+#include "internal/ptr.hh"
+
+#include <eval.hh>
+#include <globals.hh>
+#include <shared.hh>
+
+namespace nix::python {
+
+#define _public_ __attribute__((visibility("default")))
+
+PyObject * ThrownNixError = nullptr;
+PyObject * NixError = nullptr;
+
+static PyMethodDef NixMethods[] = {
+    {"callExprString", (PyCFunction) eval, METH_VARARGS | METH_KEYWORDS, "Eval nix expression"}, {NULL, NULL, 0, NULL}};
+
+static struct PyModuleDef nixmodule = {
+    PyModuleDef_HEAD_INIT, "nix", "Nix expression bindings",
+    -1, /* size of per-interpreter state of the module,
+           or -1 if the module keeps state in global variables. */
+    NixMethods};
+
+extern "C" _public_ PyObject * PyInit_nix(void)
+{
+    // These bindings can trigger builds via IFD. This means we need the build-hook to work.
+    // By default, Nix sets the build-hook to be "$(readlink /proc/self/exe) __build-remote", expecting the current
+    // binary to be Nix itself. But when we call the Nix library from Python this isn't the case, the current binary is
+    // Python then So we need to change this default, pointing it to the Nix binary instead
+    nix::settings.buildHook = nix::settings.nixBinDir + "/nix __build-remote";
+    // And by setting buildHook before calling initNix, we can override the defaults without overriding the
+    // user-provided options from the config files
+    nix::initNix();
```

Comment:
```suggestion
    nix::initLibStore();
```

A libmain dependency should not be needed anymore.

I would recommend to apply this suggestion on the cffi branch and make a new PR for that branch.

Reply:
An interrupt signal handler would be nice, but should be done the python way. I'm not sure what that is yet.
Similar for `detectStackOverflow()`, which installs a SIGSEGV handler.

These are global, cross-cutting, process-wide concerns, so we should be careful not to make assumptions.

Reply:
I'll also add that based on the way https://github.com/NixOS/rfcs/pull/136 is headed, we will most likely want multiple C binding libraries mirroring the same layering.